src/v/utils/stream_utils.h

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "random/fast_prng.h"
+#include "seastarx.h"
+#include "utils/gate_guard.h"
+#include "vassert.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/circular_buffer.hh>
+#include <seastar/core/condition-variable.hh>
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/iostream.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/semaphore.hh>
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/core/temporary_buffer.hh>
+
+#include <fmt/format.h>
+
+#include <chrono>
+#include <exception>
+#include <ratio>
+
+namespace detail {
+
+/// Helper template that wraps the input stream and consumes
+/// it to the cyclic buffer
+template<class Ch>
+class input_stream_fanout final {
+    struct data_item {
+        ss::temporary_buffer<Ch> buf;
+        ss::semaphore_units<> units;
+        unsigned mask;
+    };
+    using ring_buffer = std::deque<data_item>;
```

Comment:
Would `ss::circular_buffer` be appropriate since the size is limited to the semaphore?

Reply:
Or maybe even `ss::queue`